From 2b6c18e142b93ecfe71d91356d63018d91aa907d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Adri=C3=A1n=20Insaurralde=20Avalos?=
 <36372335+adrianinsaval@users.noreply.github.com>
Date: Fri, 9 Sep 2022 20:03:11 -0400
Subject: [PATCH 1/1] Fix incorrect handling of hookdirs config [fixes #342]

Fix `pk_alpm_config_list_options` incorrect alphabetical order, otherwise the hookdirs config is not available.

Iterate through `config->hookdirs` and use the correct API call to set additional hooks directories `alpm_option_add_hookdir`, this way SYSHOOKDIR is not overridden
---
 backends/alpm/pk-alpm-config.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/backends/alpm/pk-alpm-config.c b/backends/alpm/pk-alpm-config.c
index 0a64a1890..c96b4e167 100644
--- a/backends/alpm/pk-alpm-config.c
+++ b/backends/alpm/pk-alpm-config.c
@@ -362,6 +362,7 @@ typedef struct
 /* keep this in alphabetical order */
 static const PkAlpmConfigList pk_alpm_config_list_options[] = {
 	{ "HoldPkg", G_STRUCT_OFFSET (PkAlpmConfig, holdpkgs) },
+	{ "HookDir", G_STRUCT_OFFSET (PkAlpmConfig, hookdirs) },
 	{ "IgnoreGroup", G_STRUCT_OFFSET (PkAlpmConfig, ignoregroups) },
 	{ "IgnorePkg", G_STRUCT_OFFSET (PkAlpmConfig, ignorepkgs) },
 	{ "LocalFileSigLevel", G_STRUCT_OFFSET (PkAlpmConfig,
@@ -370,7 +371,6 @@ static const PkAlpmConfigList pk_alpm_config_list_options[] = {
 	{ "NoUpgrade", G_STRUCT_OFFSET (PkAlpmConfig, noupgrades) },
 	{ "RemoteFileSigLevel", G_STRUCT_OFFSET (PkAlpmConfig,
 						 remotefilesiglevels) },
-	{ "HookDir", G_STRUCT_OFFSET (PkAlpmConfig, hookdirs) },
 	{ NULL, 0 }
 };
 
@@ -1014,7 +1014,9 @@ pk_alpm_config_configure_alpm (PkBackend *backend, PkAlpmConfig *config, GError
 	config->noupgrades = NULL;
 
 	/* alpm takes ownership */
-	alpm_option_set_noupgrades (handle, config->hookdirs);
+	for(alpm_list_t *j = config->hookdirs; j; j = j->next) {
+		alpm_option_add_hookdir(handle, j->data);
+	}
 	config->hookdirs = NULL;
 
 	pk_alpm_config_configure_repos (backend, config, handle, error);
-- 
2.38.0

